Sources/Soto/Extensions/S3/S3+presignedPost.swift

```diff
@@ -0,0 +1,194 @@
+    public func generatePresignedPost(key: String, bucket: String, fields: [String: String] = [:], conditions: [PostPolicyCondition] = [], expiresIn: TimeInterval, now: Date = Date.now) async throws -> PresignedPostResponse {
+        // Copy the fields and conditions to a variable
+        var fields = fields
+        var conditions = conditions
+
+        // Update endpoint URL to include the bucket
+        guard let url = URL(string: endpoint) else {
+            throw S3ErrorType.presignedPost.malformedEndpointURL
+        }
+
+        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            throw S3ErrorType.presignedPost.malformedEndpointURL
+        }
+
+        guard let host = components.host else {
+            throw S3ErrorType.presignedPost.malformedEndpointURL
+        }
+
+        components.host = "\(bucket).\(host)"
+
+        guard let url = components.url else {
+            throw S3ErrorType.presignedPost.malformedBucketURL
+        }
+
+        // Gather canonical values
+        let algorithm = "AWS4-HMAC-SHA256" // Get signature version from client?
+
+        let date = now
+        let longDate = longDateFormat(date: date)
+        let shortDate = shortDateFormat(date: date)
+
+        let credential = try await getCredential(date: shortDate)
+
+        var keyCondition: PostPolicyCondition
+        let suffix = "${filename}"
+        if key.hasSuffix(suffix) {
+            keyCondition = .rule("starts-with", "$key", String(key.dropLast(suffix.count)))
+        } else {
+            keyCondition = .match("key", key)
+        }
+
+        // Add required conditions
+        conditions.append(.match("bucket", bucket))
+        conditions.append(keyCondition)
+        conditions.append(.match("x-amz-algorithm", algorithm))
+        conditions.append(.match("x-amz-date", longDate))
+        conditions.append(.match("x-amz-credential", credential))
+
+        // Add required fields
+        fields["key"] = key
+        fields["x-amz-algorithm"] = algorithm
+        fields["x-amz-date"] = longDate
+        fields["x-amz-credential"] = credential
+
+        // Create the policy and add to fields
+        let policy = PostPolicy(expiration: date.addingTimeInterval(expiresIn), conditions: conditions)
+        let stringToSign = try policy.stringToSign()
+
+        fields["Policy"] = stringToSign
+
+        // Create the signature and add to fields
+        let signature = try await getSignature(policy: stringToSign, date: shortDate)
+        fields["x-amz-signature"] = signature
+
+        // Create the response
+        let presignedPostResponse = PresignedPostResponse(url: url, fields: fields)
+
+        return presignedPostResponse
+    }
+
+    private func signingKey(date: String) async throws -> SymmetricKey {
+        let credentials = try await client.getCredential()
```

Comment:
It looks like you call `client.getCredential` twice. once here and once in `getCredential`. This can be an expensive function, you should probably reuse the value

Reply:
I'm going to unresolved this until the fixed code is submitted